### PR TITLE
Adding Test Coverage to Positions Math

### PR DESCRIPTION
**Pull request by pwdel**
* When switching shares from NO -> YES or YES -> NO, there needs to be:

1. Force selling opposite shares held.
2. Must be a negative / selling transaction showing up on record (actual sell function used).
3. Test coverage

This specific branch is building toward that whole fix, working on adding test coverage to the positions functionality which is used within selling.

Updating this MR specifically:

Ideally, we should implement banker's rounding to accurately reflect our math documentation.
We need to deal with the divide by zero / infinity case in the situation where all shares from a market have been sold and WPAM goes back to 0.5, but the normalization factor gets calculated as infinity.
We should probably implement the scenarios laid out in our mathematical documentation in terms of trades and subsequent calculations to be more demonstrative. Hypothetically if further problems are found in the software, we can add additional test cases.
Payout of initial market subsidy after final resolution can be done in a different ticket (placing in v0.0.4).

### CODE REVIEW

**Comment by pwdel** [comment]
Tests added are passing but not necessarily good at this point...I would like to go back in to the math document so I can use this to make sure the tests are sufficiently robust and make sure the math documentation matches our functionality.

**Comment by pwdel** [comment]
To Do:

1. Ideally, we should implement banker's rounding to accurately reflect our math documentation.
2. We need to deal with the divide by zero / infinity case in the situation where all shares from a market have been sold and WPAM goes back to 0.5, but the normalization factor gets calculated as infinity.
3. We should probably implement the scenarios laid out in our mathematical documentation in terms of trades and subsequent calculations to be more demonstrative. Hypothetically if further problems are found in the software, we can add additional test cases.
4. Payout of initial market subsidy after final resolution can be done in a different ticket (placing in v0.0.4).

**Comment by pwdel** [comment]
1. We'll do banker's rounding in a different ticket. https://github.com/orgs/openpredictionmarkets/projects/4/views/1?pane=issue&itemId=87892514
2. Need infinity avoidance in normalization calculation.

File: backend/handlers/math/outcomes/dbpm/marketshares_test.go [hunk 1]
```diff
@@ -1,551 +1,709 @@
...
-					continue
-				}
-				if resultPos.YesSharesOwned != expectedPos.YesSharesOwned || resultPos.NoSharesOwned != expectedPos.NoSharesOwned {
-					t.Errorf("Test %s failed for %s: expected %+v, got %+v", tc.Name, username, expectedPos, resultPos)
+			for i, payout := range actualPayouts {
+				expected := tc.ExpectedPayouts[i]
+
+				// Log both expected and actual for each mismatch
+				if payout.Payout != expected.Payout || payout.Outcome != expected.Outcome {
+					t.Errorf(
```
Comment (ajlacey) [inline backend/handlers/math/outcomes/dbpm/marketshares_test.go:222]: This will always print when running tests, it may be cleaner to just have the debug output in the error statement now that this test is written.

File: backend/handlers/math/outcomes/dbpm/marketshares_test.go [hunk 2]
```diff
@@ -1,551 +1,709 @@
...
+
+			if actualF_YES != tc.ExpectedF_YES || actualF_NO != tc.ExpectedF_NO {
+				t.Errorf(
+					"%s: expected F_YES=%f, F_NO=%f; got F_YES=%f, F_NO=%f",
+					tc.Name, tc.ExpectedF_YES, tc.ExpectedF_NO, actualF_YES, actualF_NO,
+				)
 			}
 		})
 	}
+
```
Comment (ajlacey) [inline backend/handlers/math/outcomes/dbpm/marketshares_test.go:310]: This will also lead to excessive prints when running the entire test suite

File: backend/handlers/math/outcomes/dbpm/marketshares_test.go [hunk 3]
```diff
@@ -1,551 +1,709 @@
...
 }
 
-func TestCalculateCoursePayoutsDBPM(t *testing.T) {
+func TestCalculateScaledPayoutsDBPM(t *testing.T) {
 	testcases := []struct {
-		Name               string
-		Bets               []models.Bet
-		ProbabilityChanges []wpam.ProbabilityChange
-		CoursePayouts      []CourseBetPayout
+		Name                  string
```
Comment (ajlacey) [inline backend/handlers/math/outcomes/dbpm/marketshares_test.go:315]: we don't need to include `tc.Name` because the error message is output underneath TestFunc/TestCaseName

File: backend/handlers/math/outcomes/dbpm/marketshares_test.go [hunk 4]
```diff
@@ -1,551 +1,709 @@
...
 
-func TestCalculateNormalizationFactorsDBPM(t *testing.T) {
+func TestCalculateExcess(t *testing.T) {
 	testcases := []struct {
-		Name          string
-		F_YES         float64
-		F_NO          float64
-		ExpectedF_YES float64
-		ExpectedF_NO  float64
+		Name           string
```
Comment (ajlacey) [inline backend/handlers/math/outcomes/dbpm/marketshares_test.go:408]: `"Payouts[%d]: got %d, want %d", i, payout, tc.ExpectedScaledPayouts[i]"`
```suggestion
					    "Payouts[%d]: got %d, want %d", i, payout, tc.ExpectedScaledPayouts[i]"
```

File: backend/handlers/math/outcomes/dbpm/marketshares_test.go [hunk 5]
```diff
@@ -1,551 +1,709 @@
...
-func TestAdjustPayoutsFromNewest(t *testing.T) {
+// helper function to create wpam.ProbabilityChange points succiently
+func generateProbability(probabilities ...float64) []wpam.ProbabilityChange {
+	var changes []wpam.ProbabilityChange
+	for _, p := range probabilities {
+		changes = append(changes, wpam.ProbabilityChange{Probability: p})
+	}
+	return changes
+}
+
```
Comment (ajlacey) [inline backend/handlers/math/outcomes/dbpm/marketshares_test.go:24]: I think this is a good candidate to include in our modelstesting package. We would only ever generate a bet this way when we're testing, so it would be good to make it available there.

File: backend/handlers/positions/positionsmath_test.go [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+					Outcome:  "YES",
+					PlacedAt: time.Now().Add(time.Minute),
+				},
+			},
+			ExpectedPositions: []dbpm.MarketPosition{
+				{Username: "one", NoSharesOwned: 25, YesSharesOwned: 0},
+				{Username: "two", NoSharesOwned: 0, YesSharesOwned: 5},
+			},
+		},
+		{
```
Comment (ajlacey) [inline backend/handlers/positions/positionsmath_test.go:63]: Test Failure:

expected position {Username:user1 NoSharesOwned:0 YesSharesOwned:3}, got {Username:user1 NoSharesOwned:0 YesSharesOwned:4}
expected position {Username:user2 NoSharesOwned:2 YesSharesOwned:0}, got {Username:user2 NoSharesOwned:1 YesSharesOwned:0}

File: backend/handlers/positions/positionsmath_test.go
Comment (ajlacey) [inline backend/handlers/positions/positionsmath_test.go:1]: I'm getting failures on each of these tests, is it just me?

File: backend/util/testhelpers.go
Comment (ajlacey) [inline backend/util/testhelpers.go:1]: The contents of this file are better served as a member of the `modelstesting` package given that the offset is unique to bet creation for testing purposes.

Also: "Utils" files are considered bad (tl;dr- utils is ambiguous- like misc or stuff)

**Review by ajlacey — CHANGES_REQUESTED** [review]

File: backend/util/testhelpers.go
Comment (pwdel) [reply backend/util/testhelpers.go:1]: I see what you mean here on this document.

https://google.github.io/styleguide/go/decisions

File: backend/handlers/positions/positionsmath_test.go
Comment (pwdel) [reply backend/handlers/positions/positionsmath_test.go:1]: This is what I'm getting, from the backend:

```
~/PP/socialpredict/backend [fix/positions-test-coverage] [$]
$ go test ./...
?   	socialpredict	[no test files]
?   	socialpredict/handlers	[no test files]
?   	socialpredict/handlers/admin	[no test files]
ok  	socialpredict/errors	(cached)
ok  	socialpredict/handlers/bets	(cached)
?   	socialpredict/handlers/marketpublicresponse	[no test files]
?   	socialpredict/handlers/math/financials	[no test files]
?   	socialpredict/handlers/users	[no test files]
?   	socialpredict/handlers/users/privateuser	[no test files]
?   	socialpredict/handlers/users/userhelpers	[no test files]
?   	socialpredict/logging	[no test files]
?   	socialpredict/middleware	[no test files]
?   	socialpredict/migration	[no test files]
?   	socialpredict/seed	[no test files]
?   	socialpredict/server	[no test files]
?   	socialpredict/setup/setuptesting	[no test files]
?   	socialpredict/setup	[no test files]
?   	socialpredict/util	[no test files]
ok  	socialpredict/handlers/bets/betutils	1.076s
ok  	socialpredict/handlers/markets	(cached)
ok  	socialpredict/handlers/math/market	(cached)
ok  	socialpredict/handlers/math/outcomes/dbpm	0.542s
ok  	socialpredict/handlers/math/probabilities/wpam	(cached)
ok  	socialpredict/handlers/positions	1.322s
ok  	socialpredict/handlers/setup	(cached)
ok  	socialpredict/handlers/stats	(cached) [no tests to run]
ok  	socialpredict/handlers/tradingdata	1.737s
ok  	socialpredict/handlers/users/credit	1.942s
ok  	socialpredict/handlers/users/publicuser	2.362s
ok  	socialpredict/models	(cached)
ok  	socialpredict/models/modelstesting	2.520s
```

Also, I think the pipeline is passing.

File: backend/handlers/positions/positionsmath_test.go (hunk 6 above)
Comment (pwdel) [reply backend/handlers/positions/positionsmath_test.go:63]: This appears to have been fixed/addressed.

File: backend/handlers/math/outcomes/dbpm/marketshares_test.go [hunk 7]
```diff
@@ -1,551 +1,709 @@
...
 		{
-			Name: "PreventSimultaneousSharesHeld",
+			Name:               "InitialMarketState",
+			Bets:               []models.Bet{},
+			ProbabilityChanges: generateProbability(0.500),
+			S_YES:              0,
+			S_NO:               0,
+		},
+		{
+			Name: "FirstBetNoDirection",
```
[the remark this answers is not in the document]
Comment (pwdel) [reply backend/handlers/math/outcomes/dbpm/marketshares_test.go:48]: I need to get rid of underscores anyway... I read that underscores are discouraged in golang except for the names of test files.

Basically it's the result of dividing up the market pool shares, S_ was supposed to represent S_NO and S_YES shares, the top level division.

https://github.com/openpredictionmarkets/socialpredict/blob/doc/math-latex-graphical-overviews/README/MATH/Prediction_Market_Math.pdf

File: backend/handlers/math/outcomes/dbpm/marketshares_test.go (hunk 4 above)
Comment (pwdel) [reply backend/handlers/math/outcomes/dbpm/marketshares_test.go:399]: I don't understand what this is.

File: backend/handlers/math/outcomes/dbpm/marketshares.go [hunk 8]
```diff
@@ -163,42 +147,56 @@ func CalculateScaledPayoutsDBPM(allBetsOnMarket []models.Bet, coursePayouts []Co
...
+			excess += 1           // Increment excess
 			if excess == 0 {
 				break
 			}
 		}
 	}
+	return scaledPayouts
+}
+
+// Adjust payouts to ensure they align with the available pool
```
[the remark this answers is not in the document]
Comment (pwdel) [reply backend/handlers/math/outcomes/dbpm/marketshares.go:191]: Thanks, changed but I tweaked your suggestion to:

```
// AdjustPayouts reconciles the additional or lacking funds from the betting pool by adjusting the payouts to past bets
```
This doesn't only apply to, "winners," it applies to any point in the market history, not just at the end after resolution.

During the market, it allocates excess to those participating in the market. At the end of the market, it allocates excess to those who were left over in the payout pool (winners).

File: backend/handlers/math/outcomes/dbpm/marketshares_test.go [hunk 9]
```diff
@@ -1,551 +1,709 @@
...
 	"socialpredict/handlers/math/probabilities/wpam"
 	"socialpredict/models"
 	"socialpredict/setup"
+	"socialpredict/util"
+	"sort"
 	"testing"
 	"time"
 )
 
 var now = time.Now()
```
[the remark this answers is not in the document]
Comment (pwdel) [reply backend/handlers/math/outcomes/dbpm/marketshares_test.go:15]: Thank you

File: backend/handlers/math/outcomes/dbpm/marketshares.go [hunk 10]
```diff
@@ -163,42 +147,56 @@ func CalculateScaledPayoutsDBPM(allBetsOnMarket []models.Bet, coursePayouts []Co
...
 	availablePool := marketmath.GetMarketVolume(bets)
+	return sumScaledPayouts - availablePool
+}
 
-	// Determine the excess amount
-	excess := sumScaledPayouts - availablePool
-
-	// Loop to deduct from newest to oldest until there's no excess
+// Adjust scaled payouts if excess is greater than 0
+// This  should not be possible given how the preceeding pipeline works, but we adjust for it anyway.
```
[the remark this answers is not in the document]
Comment (pwdel) [reply backend/handlers/math/outcomes/dbpm/marketshares.go:162]: I'm worried that I'm not accounting for some odd edge case, game set up or otherwise that would result in a negative excess, which would eliminate money from the system in a way we're not accounting for.  Basically this is a safeguard.

File: backend/handlers/math/outcomes/dbpm/marketshares.go [hunk 11]
```diff
@@ -163,42 +147,56 @@ func CalculateScaledPayoutsDBPM(allBetsOnMarket []models.Bet, coursePayouts []Co
...
 					break
 				}
 			}
 		}
 	}
+	return scaledPayouts
+}
 
-	// Loop to add from oldest to newest until there's no excess
+// Adjust scaled payouts if excess is less than 0
```
[the remark this answers is not in the document]
Comment (pwdel) [reply backend/handlers/math/outcomes/dbpm/marketshares.go:178]: thanks

File: backend/handlers/math/outcomes/dbpm/marketshares_test.go [hunk 12]
```diff
@@ -1,551 +1,719 @@
...
+			ScaledPayouts:  []int64{10, 20, 30},
+			Excess:         0,
+			ExpectedResult: []int64{10, 20, 30}, // No adjustment needed
+		},
+		{
+			Name:           "SmallExcess",
+			ScaledPayouts:  []int64{10, 20, 30},
+			Excess:         2,
+			ExpectedResult: []int64{10, 19, 29},
 		},
```
[the remark this answers is not in the document]
Comment (pwdel) [reply backend/handlers/math/outcomes/dbpm/marketshares_test.go:488]: Ok, added a test for 1000, now it's taking forever to run the tests, so I'm going to have to go back on what I just said in a recent comment about not being able to figure out the excess allocation logic because evidently even just dealing with 1000 is untenable.

File: backend/handlers/math/outcomes/dbpm/marketshares.go [hunk 13]
```diff
@@ -160,35 +162,62 @@ func calculateExcess(bets []models.Bet, scaledPayouts []int64) int64 {
...
+	for betIndex := range scaledPayouts {
+		scaledPayouts[betIndex] -= baseReduction
+	}
+
+	// Apply the remainder reduction to the newest bets
+	for betIndex := int64(len(scaledPayouts)) - 1; remainderReduction > 0; betIndex-- {
+		scaledPayouts[betIndex] -= 1
+		remainderReduction--
 	}
+
```
Comment (ajlacey) [inline backend/handlers/math/outcomes/dbpm/marketshares.go:188]: Nicely done! I think this should improve our runtime efficiency a good bit too!

File: backend/handlers/math/outcomes/dbpm/marketshares.go [hunk 14]
```diff
@@ -261,22 +290,12 @@ func NetAggregateMarketPositions(positions []MarketPosition) []MarketPosition {
...
-		if bet.Outcome == "YES" {
-			total += bet.Amount
-		} else if bet.Outcome == "NO" {
-			total -= bet.Amount
-		}
+// singleShareYesNoAllocator determines the outcome of a single bet.
+// Logs a fatal error and exits if the input condition (len(bets) == 1) is not met.
+func singleShareYesNoAllocator(bets []models.Bet, logger logging.Logger) string {
+	if len(bets) != 1 {
+		logger.Fatalf("singleShareYesNoAllocator: expected len(bets) = 1, got %d", len(bets))
```
Comment (ajlacey) [inline backend/handlers/math/outcomes/dbpm/marketshares.go:297]: **Will crash application**

While we do want to know if this condition occurs, I think it would be much better for it to be an error as this implementation exits the application. I think we just want to log this happened so we can inspect the conditions that caused it (maybe include what the bets are, market they're associated with, or anything else to help debug this condition).

Comment (ajlacey) [reply backend/handlers/math/outcomes/dbpm/marketshares.go:297]: That said, it's a private function, so maybe this is okay? It just could cause spurious errors to anyone building on top of this if our implementations aren't solid, or change in some unexpected way.

File: backend/models/modelstesting/testhelpers.go [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+// helper function to create wpam.ProbabilityChange points succinctly
+func GenerateProbability(probabilities ...float64) []wpam.ProbabilityChange {
+	var changes []wpam.ProbabilityChange
+	for _, p := range probabilities {
+		changes = append(changes, wpam.ProbabilityChange{Probability: p})
+	}
+	return changes
+}
```
Comment (ajlacey) [inline backend/models/modelstesting/testhelpers.go:28]: This function is not related to the models in any way. It should be added to a `wpam/wpamtesting` module that can be used anywhere the `wpam` module is already available.

**Review by ajlacey — APPROVED** [review]
Nicely done, this is really close to finished so I'm going to approve it. I made a couple of comments for your consideration before merging. The two areas I think are most important are: 
- Do we want to exit the entire application if the number of bets is unexpected (i.e. not equal to 1), or should we just log an error instead?
- Can we move the wpam probabilities helper into a new module to support testing of wpam?


Also, just want to say really nice work on this PR, I feel like a lot of variable names are a lot easier to understand now, and the code flow is looking a lot more standard. Thanks for putting in the effort here 💯

File: backend/handlers/math/outcomes/dbpm/marketshares.go (hunk 14 above)
Comment (pwdel) [reply backend/handlers/math/outcomes/dbpm/marketshares.go:297]: I put together a variety of comments on this embedded as responses as I was building it out. Ultimately I think we need to have a discussion about whether if something breaks the economy, does that break the app?

https://github.com/orgs/openpredictionmarkets/projects/9/views/1?pane=issue&itemId=88863680&issue=openpredictionmarkets%7Csocialpredict%7C412

So if we were building nuclear warhead launching software, we would want any error to go to fatal, because we only want the warhead to launch as intended. Likewise, we need to decide that if the economy gets wrecked due to a programming oversight, does that wreck everything we're trying to do? If real money ever gets attached to SocialPredict, it seems that we should be taking a conservative approach.
